src/observersTools/leggedOdometryTools.cpp

```diff
@@ -0,0 +1,860 @@
+  {
+    // the kinematics of the contacts are the ones of the surface, but we must transport the measured wrench
+    sva::PTransformd worldSurfacePoseOdometryRobot = odometryRobot().surfacePose(contact.surfaceName());
+    contact.currentWorldKine_ =
+        kinematicsTools::poseFromSva(worldSurfacePoseOdometryRobot, so::kine::Kinematics::Flags::pose);
+
+    so::kine::Kinematics contactSensorKine = contact.currentWorldKine_.getInverse() * worldSensorKineOdometryRobot;
+    // expressing the force measurement in the frame of the surface
+    contact.forceNorm_ = (contactSensorKine.orientation * fs.wrenchWithoutGravity(odometryRobot()).force()).norm();
+  }
+
+  return contact.currentWorldKine_;
+}
+
+void LeggedOdometryManager::selectForOrientationOdometry()
+{
+  contactsManager_.oriOdometryContacts_.clear();
+  for(auto it = contactsManager_.contactsFound().begin(); it != contactsManager_.contactsFound().end(); it++)
+  {
+    LoContactWithSensor & contact = contactsManager_.contactWithSensor(*it);
+    if(contact.getName().find("Hand") == std::string::npos
+       && contact.wasAlreadySet_) // we don't use hands for the orientation odometry
+    {
+      contact.useForOrientation_ = true;
+      contactsManager_.oriOdometryContacts_.insert(contact);
+    }
+  }
+  // contacts are sorted from the lowest force to the highest force
+  while(contactsManager_.oriOdometryContacts_.size() > 2)
+  {
+    (*contactsManager_.oriOdometryContacts_.begin()).get().useForOrientation_ = false;
+    contactsManager_.oriOdometryContacts_.erase(contactsManager_.oriOdometryContacts_.begin());
+  }
+}
+
+void LeggedOdometryManager::addContactLogEntries(mc_rtc::Logger & logger, const LoContactWithSensor & contact)
+{
+  const std::string & contactName = contact.getName();
+  kinematicsTools::addToLogger(contact.worldRefKine_, logger, odometryName_ + "_" + contactName + "_refPose");
+  kinematicsTools::addToLogger(contact.currentWorldFbPose_, logger,
+                               odometryName_ + "_" + contactName + "_currentWorldFbPose");
+  kinematicsTools::addToLogger(contact.currentWorldKine_, logger,
+                               odometryName_ + "_" + contactName + "_currentWorldContactKine");
+}
+
+void LeggedOdometryManager::removeContactLogEntries(mc_rtc::Logger & logger, const LoContactWithSensor & contact)
+{
+  const std::string & contactName = contact.getName();
+  logger.removeLogEntry(odometryName_ + "_" + contactName + "_ref_position");
+  logger.removeLogEntry(odometryName_ + "_" + contactName + "_ref_orientation");
+  kinematicsTools::removeFromLogger(logger, odometryName_ + "_" + contactName + "_refPose");
+  kinematicsTools::removeFromLogger(logger, odometryName_ + "_" + contactName + "_currentWorldFbPose");
+  kinematicsTools::removeFromLogger(logger, odometryName_ + "_" + contactName + "_currentWorldContactKine");
+}
+
+so::kine::Kinematics & LeggedOdometryManager::getAnchorFramePose(const mc_control::MCController & ctl)
```

Comment:
I haven't looked too closely but I guess this can be implemented by calling `getAnchorFramePose(ctl, "")` and checking in the other implementation if the sensor is empty or not before accessing it?